keps/sig-api-machinery/3352-aggregated-discovery/README.md

```diff
@@ -0,0 +1,874 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!-- This section is incredibly important for producing high-quality,
+user-focused documentation such as release notes or a development
+roadmap. It should be possible to collect this information before
+implementation begins, in order to avoid requiring implementors to
+split their attention between writing release notes and implementing
+the feature itself. KEP editors and SIG Docs should help to ensure
+that the tone and content of the `Summary` section is useful for a
+wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the
+[documentation style guide]. In particular, wrap lines to a reasonable
+length, to make it easier for reviewers to cite specific portions, and
+to minimize diff churn on updates.
+-->
+
+
+The operations that a Kubernetes API server supports are reported
+through a collection of small documents partitioned by group-version.
+All clients of Kubernetes APIs must send a request to every
+group-version in order to "discover" the available APIs. This causes a
+storm of requests for clusters and is a source of latency and
+throttling. This KEP proposes centralizing the "discovery" mechanism
+into one document so that clients only need to make one request to the
+API server to retrieve all the operations available.
+
+## Motivation
+
+<!-- This section is for explicitly listing the motivation, goals, and
+non-goals of this KEP. Describe why the change is important and the
+benefits to users. The motivation section can optionally provide links
+to [experience reports] to demonstrate the interest in a KEP within
+the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports -->
+
+All clients and users of Kubernetes APIs usually first need to
+“discover” what the available APIs are and how they can be used. These
+APIs are described through a mechanism called “Discovery” which is
+typically queried to then build the requests to correct APIs.
+Unfortunately, the “Discovery” API is made of lots of small objects
+that need to be queried individually, causing possibly a lot of delay
+due to the latency of each individual request (up to 80 requests, with
+most objects being less than 1,024 bytes). The more numerous the APIs
+provided by the Kubernetes cluster, the more requests need to be
+performed.
+
+The most well known Kubernetes client that uses the discovery
+mechanism is `kubectl`, and more specifically the
+`CachedDiscoveryClient` in `client-go`. To mitigate some of this
+latency, kubectl has implemented a 6 hour timer during which the
+discovery API is not refreshed. The drawback of this approach is that
+the freshness of the cache is doubtful and the entire discovery API
+needs to be refreshed after 6 hours, even if it hasn’t expired.
+
+This not only impacts kubectl, but all clients of kubernetes. We can
+do better.
```

Comment:
Some of the nuances we discussed in comments deserve to be in the KEP:

1. Which clients are we targeting primarily: clients that need to frequently request or maintain a discovery cache, like kubectl or web interfaces
2. We want all clients to benefit from this, but our alpha is targeted at solving these problems for those key consumers
3. We have some use cases that may require additional work to correctly support, we should identify those before entry to beta / reach consensus on their utility:
    * Namespace controller needs to ensure that discovery documents are refreshed after the namespace goes into the terminating phase (needs to guarantee that it sees the list of resources that could have been created within the namespace before the namespace is deleted).  The controller needs to be able to guarantee discovery doc list happens-after that, which is best solved by letting the namespace controller "wait" for the next refresh.  What we can't allow is for the namespace controller to not observe that resource
     * The use of server side caching may break *some* client side batch loops done with kubectl that need happens-before semantics, we should consider that usecase in kubectl and ensure we have a mitigation
4. Polling being sufficient over watch for most users
5. The API types in meta/v1 should be meta/v1alpha1 to start, not meta/v1